### PR TITLE
[SPARK-15241][SPARK-15242][SQL] fix 2 decimal-related issues in RowEncoder

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

SPARK-15241: We now support java decimal and catalyst decimal in external row, it makes sense to also support scala decimal.

SPARK-15242: This is a long-standing bug, and is exposed after https://github.com/apache/spark/pull/12364, which eliminate the `If` expression if the field is not nullable:

```
val fieldValue = serializerFor(
  GetExternalRowField(inputObject, i, externalDataTypeForInput(f.dataType)),
  f.dataType)
if (f.nullable) {
  If(
    Invoke(inputObject, "isNullAt", BooleanType, Literal(i) :: Nil),
    Literal.create(null, f.dataType),
    fieldValue)
} else {
  fieldValue
}
```

Previously, we always use `DecimalType.SYSTEM_DEFAULT` as the output type of converted decimal field, which is wrong as it doesn't match the real decimal type. However, it works well because we always put converted field into `If` expression to do the null check, and `If` use its `trueValue`'s data type as its output type.
Now if we have a not nullable decimal field, then the converted field's output type will be `DecimalType.SYSTEM_DEFAULT`, and we will write wrong data into unsafe row.

The fix is simple, just use the given decimal type as the output type of converted decimal field.

These 2 issues was found at https://github.com/apache/spark/pull/13008
## How was this patch tested?

new tests in RowEncoderSuite

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @yhuai @liancheng @viirya

**Comment by rxin** [comment]
actually cc @davies also

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala [hunk 1]
```diff
@@ -143,21 +143,35 @@ class RowEncoderSuite extends SparkFunSuite {
     assert(input.getStruct(0) == convertedBack.getStruct(0))
   }
 
-  test("encode/decode Decimal") {
+  test("encode/decode decimal type") {
     val schema = new StructType()
-      .add("int", IntegerType)
-      .add("string", StringType)
-      .add("double", DoubleType)
-      .add("decimal", DecimalType.SYSTEM_DEFAULT)
```
Comment (yhuai) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/encoders/RowEncoderSuite.scala]: I feel it is still good to keep int, string, and double columns?

**Comment by viirya** [comment]
LGTM

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master and 2.0, thanks!

**Pull request closed without merging**